src/process-win.cpp

```diff
@@ -102,28 +106,47 @@ auto Process::get_status() const -> Status {
 }
 
 auto Process::collect_outputs() -> bool {
-    const auto handles      = std::array{pipes[1].output, pipes[2].output};
-    const auto wait_process = WaitForSingleObject(process_handle, 0);
-    if(wait_process == WAIT_OBJECT_0) {
-        status = Status::Finished;
-    }
-    for(auto i = 0; i < 2; i += 1) {
-        auto buf = std::array<char, 256>();
-        while(true) {
-            auto bytes_avail = DWORD();
-            if(PeekNamedPipe(handles[i], NULL, 0, NULL, &bytes_avail, NULL) == 0 || bytes_avail == 0) {
-                break;
-            }
+    // child threads to collect the (stdout, stderr) outputs
+    auto threads = std::array<std::thread, 2>(); // stdout, stderr
+    for(auto i = 0; i < threads.size(); i += 1) {
+        threads[i] = std::thread([this, i]() {
+            auto buf = std::array<char, 256>();
             auto len = DWORD();
-            if(ReadFile(handles[i], buf.data(), buf.size(), &len, NULL) != TRUE || len <= 0) {
-                break;
+            while(true) {
+                if(ReadFile(pipes[i + 1].output, buf.data(), buf.size(), &len, NULL) != TRUE || len <= 0) {
+                    break;
+                }
+                if(status == Status::Finished) {
+                    // finished the main process
+                    break;
+                }
+                auto callback = i == 0 ? on_stdout : on_stderr;
+                if(callback) {
+                    callback({buf.data(), size_t(len)});
+                }
             }
-            auto callback = i == 0 ? on_stdout : on_stderr;
-            if(callback) {
-                callback({buf.data(), size_t(len)});
+        });
+    }
+
+    // main process
+    while(true) {
+        const auto wait_process = WaitForSingleObject(process_handle, INFINITE);
+        if(wait_process == WAIT_OBJECT_0) {
```

Comment:
add assert to wait_process == WAIT_OBJECT_0 and remove this branch.
return value other than WAIT_OBJECT_0 is unexpected.